gateway/ln-gateway/src/lib.rs

```diff
@@ -772,102 +788,106 @@ impl Gateway {
+                    // the clients for the first time, just add them to
+                    // the in-memory maps
+                    let scid = config.mint_channel_id;
+                    self.clients.write().await.insert(federation_id, client);
+                    self.scid_to_federation
+                        .write()
                         .await
-                    {
-                        // Registering each client happens in the background, since we're loading
-                        // the clients for the first time, just add them to
-                        // the in-memory maps
-                        let scid = config.mint_channel_id;
-                        self.clients.write().await.insert(federation_id, client);
-                        self.scid_to_federation
-                            .write()
-                            .await
-                            .insert(scid, federation_id);
-                    } else {
-                        warn!("Failed to load client for federation: {federation_id}");
-                    }
+                        .insert(scid, federation_id);
+                } else {
+                    warn!("Failed to load client for federation: {federation_id}");
+                }
 
-                    if config.mint_channel_id > next_channel_id {
-                        next_channel_id = config.mint_channel_id + 1;
-                    }
+                if config.mint_channel_id > next_channel_id {
+                    next_channel_id = config.mint_channel_id + 1;
                 }
-                channel_id_generator.store(next_channel_id, Ordering::SeqCst);
             }
-            return Ok(());
+            channel_id_generator.store(next_channel_id, Ordering::SeqCst);
+            Ok(())
+        } else {
+            Err(GatewayError::Disconnected)
         }
-
-        Err(GatewayError::Disconnected)
     }
 
     async fn register_clients_timer(&mut self, task_group: &mut TaskGroup) {
-        if let GatewayState::Running {
-            lnrpc,
-            lightning_public_key: _,
-            lightning_alias: _,
-        } = self.state.read().await.clone()
-        {
-            let clients = self.clients.clone();
-            let api = self.api_addr.clone();
-            let gateway_id = self.gateway_id;
-            let num_route_hints = self.num_route_hints;
-            task_group
-                .spawn("register clients", move |handle| async move {
-                    while !handle.is_shutting_down() {
+        let clients = self.clients.clone();
+        let api = self.api_addr.clone();
+        let gateway_id = self.gateway_id;
+        let num_route_hints = self.num_route_hints;
+        let state = self.state.clone();
+        task_group
+            .spawn("register clients", move |handle| async move {
+                let registration_loop = async {
+                    loop {
```

Comment:
How does this loop break? We should probably add a `is_shutting_down` check right?

Reply:
Good question! I added a comment above `tokio::select!`, but let me know if that isn't clear. I was looking for a way to receive a shutdown signal without needing to wait for the next iteration of the loop, which will include a long registration delay. Using `tokio::select` seemed like the cleanest approach, but let me know if you think there's a better way to short circuit the loop on a shutdown signal.

I didn't see any harm in allowing the task to cancel while in the middle of registering, but please let me know if you see an issue if only a subset of the clients finish registering with the federation when the task terminates.

Reply:
If there is an issue with only a subset of the clients registering before the task terminates, we can tweak this to `tokio::select` on either the shutdown signal or the registration delay, while ensuring the logic to register clients with the federation finishes before the task terminates.

Reply:
No, I don't think that's an issue. I see the gateway registration as "best effort" anyway, since its possible that a gateway registers then immediately goes offline. 

`tokio::select` should be fine too, I'm mainly wondering if the `registration_loop` future in `tokio::select` is even necessary, since it doesn't look like that loop ever breaks.

Reply:
If we don't use the registration loop in the select and instead add a check for `is_shutting_down` at the start of the loop, shutting down the task group gracefully can take up to 85% of the TTL (8.5 mins).

The existing behavior is an infinite loop to re-register clients, so this just modifies it to handle a shutdown signal gracefully without needing to finish sleeping and registering clients.